### ISSUE TITLE
Query Engine Design

### ISSUE THREAD

**Issue by benbjohnson** [influxdata/influxdb#4631]
# Query Engine

## Overview

The current state of the query engine makes it difficult to maintain and to add features. This is a proposal to refactor the engine from its current map/reduce design to a streaming design.

This proposal has the following design goals:

1. Ease development of additional statistic functions
2. Minimize allocations
3. Well-defined, performant remote serialization
4. Support joining multiple series
5. Support subqueries
6. Unify the separate raw & aggregate execution paths

## Current Model

The existing query engine works as a 2-level map/reduce. On the aggregate side, one Mapper is created for each shard and the Executor explicitly handles draining each one for a given time window and handing off the values to the Reducer. The results from the reducers are then processed for fills, derivatives, and other aggregate preprocessing before being sent back as rows to the client.

For raw queries, a 2-level map/reduce is performed, however, the raw mapper sorts the points as they're being streamed out using `container/heap`.

There are several issues with the current approach:

1. The fixed 2-level map/reduce does not allow us to support joins or subqueries

2. Points returned from the mapper can be either a primitive value or a map of keys to primative values. This breaks the type system and causes a LOT of handler code as well as a lot of memory allocations.

3. Aggregate & raw execution paths are separated because the processing in them is not broken out into encapsulated pieces.

## Proposed Model

The model that I propose reduces everything to a pull-based stream of points. These points are composed of a timestamp, a typed value (`int`, `float`, `string`, etc), and a set of tags. Since this is a pull model, these are called "iterators":

```
type Float64Iterator interface {
    Next() (timestamp int64, value float64, tags map[string]string)
}

type Int64Iterator interface {
    Next() (timestamp int64, value int64, tags map[string]string)
}

type StringIterator interface {
    Next() (timestamp int64, value string, tags map[string]string)
}

...
```

### Iterator types

The following types of iterators would exist:

* Raw data from the storage engine. Each field in a series is a separate iterator.

* InfluxQL functions. These would include the parameters for the function (interval, top/botton N, etc). Many functions output simple types (e.g. `count`, `sum`, `min`, `max`), others operate on the raw stream (`stddev`, `median`, `percentile`) while a few would need specialized iterator types (`mean` would output a `TotalCountIterator` to be reduced into a `Float64Iterator`).

* Math operations. Binary operators would take two input iterators and add/subtract/multiply/etc and produce an output iterator.

* Select fields. By definition, since raw data, functions, and math operations produce an iterator, each field in a `SELECT` statement would be an iterator. This allows us to seamlessly allow nested queries by making the fields the input iterators to another query.

Single iterators could also be multiplexed as inputs across multiple iterators for efficiency. For example, if a query uses `min(x)` and `max(x)`, the `x` iterator should be multiplexed into 2 iterators as input to the `min` and `max` iterators but the data should be read from the storage engine only once. This is outside the scope of this proposal but is a consideration in the design.

### Joins & filters

Consolidating everything down to an iterator poses two problems:

1. Values from one iterator may need to be filtered with values from other iterators.

2. Iterators are not synced in terms of timestamps so output points need to be joined on timestamp.

So another construct needs to be added: the join. A join takes `N` input iterators and returns `N` output iterators. The purpose of the join is to:

* Send a value to each of the output iterators for the next timestamp available across the input iterators. Input iterators without a point matching this timestamp will output a null value. This makes it easy to produce output rows as well as perform math operations on each row.

* Update a map for all fields in the condition and then evaluate the condition. If the condition evaluates false then the timestamp is skipped. Consolidating the condition evaluating in the join means that only 1 map needs to be allocated.

## Example plans

The following are example plans for several queries.

### Raw query with filter

```
SELECT x, y FROM m WHERE z > 10
```

```
          ┌────────────────────┐  ┌────────────────────┐            
          │   FloatIterator    │  │   FloatIterator    │            
          │        <x>         │  │        <y>         │            
          └────────────────────┘  └────────────────────┘            
                     ▲                       ▲                      
                     └────────────┬──────────┘                      
                                  │                                 
                         ┌─────────────────┐                        
                         │      Join       │                        
                         │ (filter z > 10) │                        
                         └─────────────────┘                        
                                  ▲                                 
           ┌──────────────────────┼──────────────────────┐          
           │                      │                      │          
┌────────────────────┐ ┌────────────────────┐ ┌────────────────────┐
│ tsm1.FloatIterator │ │ tsm1.FloatIterator │ │  tsm1.IntIterator  │
│       <m.x>        │ │       <m.y>        │ │       <m.z>        │
└────────────────────┘ └────────────────────┘ └────────────────────┘
```

### Aggregate query from multiple shards

```
SELECT min(x), max(y) FROM m GROUP BY time(10m)
```

```
                             ┌────────────────────┐  ┌────────────────────┐                             
                             │   FloatIterator    │  │   FloatIterator    │                             
                             │        <x>         │  │        <x>         │                             
                             └────────────────────┘  └────────────────────┘                             
                                        ▲                       ▲                                       
                                        └───────────┬───────────┘                                       
                                                    │                                                   
                                         ┌────────────────────┐                                         
                                         │        Join        │                                         
                                         └────────────────────┘                                         
                                                    ▲                                                   
                                        ┌───────────┴──────────┐                                        
                                        │                      │                                        
                             ┌────────────────────┐ ┌────────────────────┐                              
                             │    MinIterator     │ │    MaxIterator     │                              
                             │        <x>         │ │        <y>         │                              
                             └────────────────────┘ └────────────────────┘                              
                                        ▲                      ▲                                        
                                        │                      │                                        
                                     ┌──┼──────────────────────┴──────────────────────────┐             
              ┌──────────────────────┼──┴───────────────────────────┐                     │             
              │                      │                              │                     │             
              │                      │                              │                     │             
Shard #1──────┼──────────────────────┼───────────┐    Shard #2──────┼─────────────────────┼────────────┐
│             │                      │           │    │             │                     │            │
│  ┌────────────────────┐ ┌────────────────────┐ │    │  ┌────────────────────┐ ┌────────────────────┐ │
│  │    MinIterator     │ │    MaxIterator     │ │    │  │    MinIterator     │ │    MaxIterator     │ │
│  │        <x>         │ │        <y>         │ │    │  │        <x>         │ │        <y>         │ │
│  └────────────────────┘ └────────────────────┘ │    │  └────────────────────┘ └────────────────────┘ │
│             ▲                      ▲           │    │             ▲                      ▲           │
│             │                      │           │    │             │                      │           │
│  ┌────────────────────┐ ┌────────────────────┐ │    │  ┌────────────────────┐ ┌────────────────────┐ │
│  │ tsm1.FloatIterator │ │ tsm1.FloatIterator │ │    │  │ tsm1.FloatIterator │ │ tsm1.FloatIterator │ │
│  │       <m.x>        │ │       <m.y>        │ │    │  │       <m.x>        │ │       <m.y>        │ │
│  └────────────────────┘ └────────────────────┘ │    │  └────────────────────┘ └────────────────────┘ │
│                                                │    │                                                │
└────────────────────────────────────────────────┘    └────────────────────────────────────────────────┘
```

**Comment by pauldix** [comment]
How would you see this working if you have many underlying series that are getting merged into one? For example say you have this data:

```
cpu,host=A,region=west used=23.2
cpu,host=B,region=west used=95.3
```

Then when you do a query like: 

```sql
SELECT min(used) from cpu
WHERE region = 'west' AND time > now() - 4h
GROUP BY time(10m)
```

Would you end up joining the two underlying series together on time before or after performing the min on each?

When you join them together, you'd have to do time comparisons. It feels like this might resurface the problem we saw with queries that merge together thousands of series.

Is there some structure you have in mind that would avoid all those comparisons until later?

**Comment by benbjohnson** [comment]
@pauldix For a group-by `min`, we have relaxed sorting requirements. The low level iterators over the storage engine would need to know the group by interval used in the query and be required to emit all points within an interval from all series before moving to the next interval. They don't need to be sorted.

It would look like this:

```
FOREACH interval
  FOREACH cursor
    FOREACH point IN cursor
      IF point.timestamp > END
        cursor.unread(point)
        BREAK
      END

      emit(point)
    END
  END
END
```

So then for 2 sorted cursors with the following timestamps:

```
CURSOR A: 1s,3s,6s,11s,14s,21s
CURSOR B: 2s,8s,10s,13s,25s
```

An iterator with a `GROUP BY time(10s)` would output:

```
1s,3s,6s,2s,8s,10s,11s,14s,13s,21s,25s
```

The individual points aren't sorted but it does have a guarantee about intervals being sorted which a parent `MinIterator` would use to calculate the minimum for each interval.

**Comment by pauldix** [comment]
That makes sense. Since it's streaming, I assume it means we also don't need to keep all the points for an interval in memory?

If we're passing the interval information down to the low level iterators, we should make it so that intervals don't need to start at a bucketed time.

For example if you instead wanted 1m windows, but have the windows start from the time of the query (as opposed to `tsec % 60`

**Comment by benbjohnson** [comment]
@pauldix It depends on the iterator itself. Most function iterators (`min`, `max`, `count`, etc) can discard points immediately after processing them. Other ones like `stddev` and `percentile` need to collect all points in the interval. Other ones, like rolling averages could manage their own complexity of bucketing within the iterator itself.

I agree about tweaking the windowing. The iterators should have a size and an offset.

**Comment by shanielh** [comment]
So looking forward to it, Subqueries :+1:

**Comment by otoolep** [comment]
How will this design account for the fact that two shards may have different types for the same field, for a given measurement?

**Comment by otoolep** [comment]
I'd like more details on how the distributed case is going to work.

**Comment by corylanou** [comment]
For int vs float fields I suspect we'll have to throw an error.  In general, it would be nice to be able to return an error from the query engine, which currently tends to be very difficult if they error occurs at the iterator (shard) level.  If we added the ability to return errors then we could handle these types of scenarios as well

The one issue that I would like to be able to correct as well is when you use a mathematical function on a non-math type, like bool or string. Right now the query engine just fails silently so to speak by not calculating those types, but it should either error out or return a warning (I personally prefer an error in that case).

**Comment by benbjohnson** [comment]
@otoolep 

> How will this design account for the fact that two shards may have different types for the same field

If two iterators are of different types then the one should be cast to the other in a deterministic way. For example, an iterator from an older shard could set a `float` type for the field and then if it were changed to `int` in a newer shard's iterator then that iterator would cast each item to a float.

> I'd like more details on how the distributed case is going to work.

The shards will execute in isolation merged via higher level iterators. The fact that an iterator is remote is abstracted away.


@corylanou 

> For int vs float fields I suspect we'll have to throw an error

Ints and floats can be cast back and forth. Casting strings and bools to numbers would cause them to be `NaN` or null. We allow changing data types over time so I don't think an error is appropriate. Maybe a warning.

**Comment by otoolep** [comment]
>The shards will execute in isolation merged via higher level iterators. The fact that an iterator is remote is abstracted away.

Yeah, I follow that. I guess I'll just wait for the code to see how it will work. We had lots of trouble with remote mapping, the way it was coded. In the examples above, it looks like the "join" operation will take place on the co-ordinator node (the node receiving the query). Is that correct?  If so, it sounds like the "join" code will be doing a type-assertion on the iterators to determine which underlying type is being iterated (int, float etc).

I can just wait for the code, if it's easier. The main thing is that you know what I am talking about. :-)

**Comment by corylanou** [comment]
We should start sending `Request-Id` through our system.  So if we are re-architecting the query engine, it should expect a `Requeset-Id` maybe at the sql-statement level, and any time it logs anything will reference that ID.

Here is the issue: https://github.com/influxdb/influxdb/issues/4637

**Comment by otoolep** [comment]
@benbjohnson - can you confirm the new design will easily support queries such as the following:

`SELECT foo * 2 FROM bar`
`SELECT foo * foo FROM bar`
`SELECT foo / qux FROM bar`

**Comment by benbjohnson** [comment]
@otoolep Yes, math will be supported through a `BinOpIterator` that joins `LHS` & `RHS` into a single point. They can be composed for more complicated expressions as well.

**Comment by otoolep** [comment]
How many records will be pulled from the disk if `LIMIT` is set? Right now there is a bug that way too many records are pulled from `LIMIT 1` and results in poor performance.

**Comment by benbjohnson** [comment]
@otoolep Once the first row is computed and returned, you can close the iterators and it should stop reading from disk. You may get some extra disk reads due to buffering ahead though.

**Comment by jsternberg** [comment]
Closing this since the new query engine was merged and has been in use since 0.11.
